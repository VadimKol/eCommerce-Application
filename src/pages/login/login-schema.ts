import { z } from 'zod';

export const loginSchema = z.object({
  email: z.string().email('Email addresses must be properly formatted (e.g., user@example.com).'),
  password: z
    .string()
    .min(8, 'Minimum 8 characters')
    .regex(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .regex(/[a-z]/, 'Password must contain at least one lowercase letter')
    .regex(/[0-9]/, 'Password must contain at least one digit')
    .refine(
      (val) => val[0] !== ' ' && val[val.length - 1] !== ' ',
      'Password must not contain leading or trailing whitespace.',
    ),
});

export type LoginSchema = z.infer<typeof loginSchema>;
