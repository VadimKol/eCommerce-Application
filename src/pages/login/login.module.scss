.main {
  width: calc(100% - 40px);
  margin: 0 20px;
}

.email_label,
.password_label {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 20px;
}

.login {
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  max-width: 400px;
  margin: auto;
  padding: 20px 40px 10px;
  color: var(--primary-color);
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(45deg, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;

  &_title {
    margin-top: 0;
    text-align: center;
  }

  &_button {
    width: 60%;
    height: 40px;
    margin: auto;
    font-size: 16px;
    color: var(--text-color);
    background: var(--primary-color);
    border: 1px solid var(--primary-color);
    border-radius: 6px;
    transition: all 500ms;

    &:hover {
      box-shadow: -5px 5px 5px var(--secondary-color);
      cursor: pointer;
      transition: all 500ms;
      transform: translateY(-2px);
    }
  }

  & .email {
    padding: 0 5px;
  }

  & .password {
    padding: 0 30px 0 5px;
  }

  & .email,
  & .password {
    box-sizing: border-box;
    width: 73%;
    height: 30px;
    border: 1px solid var(--primary-color);
    border-radius: 6px;

    &:focus {
      border: 1px solid #000000;
    }

    &:focus-visible {
      outline: none;
    }
  }

  & .valid {
    box-shadow: 0 0 7px var(--secondary-color);
  }

  & .invalid {
    box-shadow: 0 0 7px var(--error-color);
  }

  & .error {
    width: 100%;
    margin-bottom: 20px;
    color: var(--error-color);
  }

  & .password_block {
    position: relative;
  }

  & .reveal {
    position: absolute;
    top: 6px;
    right: 6px;
    width: 20px;
    height: 20px;
    background-color: transparent;
    background-image: url("/src/assets/images/hidden.svg");
    background-size: cover;
    border: none;


    @media (hover: hover) {
      &:hover {
        cursor: pointer;
      }
    }
  }

  & .show {
    background-image: url("/src/assets/images/show.svg");
  }

  @media screen and (width <= 440px) {
    padding: 20px 20px 10px;
  }
}

.registry_link {
  margin-top: 30px;
  text-align: center;
}

.link {
  font-weight: 600;
  text-decoration: none;
  color: var(--secondary-color);
}

.disabled {
  background-color: var(--disabled-color);
  border-color: var(--disabled-color);
  pointer-events: none;
  user-select: none;
}
