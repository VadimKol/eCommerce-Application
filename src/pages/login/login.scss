.email-label,
.password-label {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 20px;
}

.login {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  max-width: 400px;
  height: fit-content;
  margin: auto;
  padding: 20px 40px 10px;
  color: #00215e;
  background:
    linear-gradient(#ffffff 0 0) padding-box,
    linear-gradient(45deg, #00215e, #0ed145) border-box;
  border: 5px solid transparent;
  border-radius: 10px;

  &-title {
    margin-top: 0;
    text-align: center;
  }

  &-button {
    width: 60%;
    height: 40px;
    margin: auto;
    font-size: 16px;
    color: #ffffff;
    background: #00215e;
    border: 1px solid #00215e;
    border-radius: 6px;
    transition: all 500ms;

    &:hover {
      box-shadow: -5px 5px 5px #00215e;
      cursor: pointer;
      transition: all 500ms;
      transform: translateY(-2px);
    }
  }

  & .email {
    padding: 0 5px;
  }

  & .password {
    padding: 0 30px 0 5px;
  }

  & .email,
  & .password {
    box-sizing: border-box;
    width: 73%;
    height: 30px;
    border: 1px solid #00215e;
    border-radius: 6px;

    &:focus {
      border: 1px solid #000000;
    }

    &:focus-visible {
      outline: none;
    }
  }

  & .valid {
    box-shadow: 0 0 7px #0ed145;
  }

  & .invalid {
    box-shadow: 0 0 7px #ff0000;
  }

  & .error {
    width: 100%;
    margin-bottom: 20px;
    color: #ff0000;
  }

  & .password-block {
    position: relative;
  }

  & .reveal {
    position: absolute;
    top: 6px;
    right: 6px;
    width: 20px;
    height: 20px;
    background-color: transparent;
    background-image: url("/src/assets/images/hidden.svg");
    background-size: cover;
    border: none;


    @media (hover: hover) {
      &:hover {
        cursor: pointer;
      }
    }
  }

  & .show {
    background-image: url("/src/assets/images/show.svg");
  }

  @media screen and (width <= 440px) {
    margin: auto 20px;
    padding: 20px 20px 10px;
  }
}

.registry-link {
  margin-top: 30px;
  text-align: center;
}

.disabled {
  background-color: #808080;
  border-color: #808080;
  pointer-events: none;
  user-select: none;
}
