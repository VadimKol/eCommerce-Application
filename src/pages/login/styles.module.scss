.login {
  box-sizing: border-box;
}
