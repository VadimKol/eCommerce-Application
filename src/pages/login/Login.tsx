import { zodResolver } from '@hookform/resolvers/zod';
import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { Link } from 'react-router-dom';
import { toast } from 'react-toastify';

import { deleteCart } from '@/api/cart';
import { login } from '@/api/client-actions';
import { ActionPaths } from '@/common/enums';
import { broadcastChannel } from '@/common/utils';
import { CustomButton } from '@/components/custom-button/СustomButton';
import { useAuth } from '@/hooks/useAuth';
import { useCart } from '@/hooks/useCart';

import { type LoginSchema, loginSchema } from './login-schema';
import styles from './styles.module.scss';

export function Login(): JSX.Element {
  const { handleLogin } = useAuth();
  const [revealPassword, setRevealPassword] = useState(false);
  const {
    register,
    getFieldState,
    watch,
    formState: { errors, isValid },
  } = useForm<LoginSchema>({ mode: 'onChange', resolver: zodResolver(loginSchema) });
  const { onChange: onChangeEmail, name: Email, ref: refEmail } = register('email');
  const { onChange: onChangePassword, name: Password, ref: refPassword } = register('password');
  const emailState = getFieldState('email');
  const passwordState = getFieldState('password');
  const email = watch('email');
  const password = watch('password');
  const { cart, updateCart } = useCart();

  let emailClass = styles.email;
  let passwordClass = styles.password;

  if (emailState.isDirty) {
    emailClass += emailState.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  if (passwordState.isDirty) {
    passwordClass += passwordState.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  return (
    <main className="main">
      <form
        className={styles.login}
        onSubmit={(event) => {
          event.preventDefault();
          if (isValid) {
            login({
              email,
              password,
              anonymousCart: { typeId: 'cart', id: cart?.id },
              anonymousId: cart?.anonymousId,
              anonymousCartSignInMode: 'MergeWithExistingCustomerCart',
            })
              .then(async (response) => {
                localStorage.setItem('geek-shop-auth', 'true');
                toast(`Hello ${response.body.customer.firstName}`, { type: 'success' });
                handleLogin();
                updateCart(response.body.cart || null);
                broadcastChannel.postMessage({ type: 'Auth', payload: { cart: response.body.cart } });
                try {
                  await deleteCart(cart?.id || '');
                } catch {
                  throw new Error(`Failed to delete cart`);
                }
              })
              .catch((error: Error) => toast(error.message, { type: 'error' }));
          } else {
            toast('Validation error', { type: 'error' });
          }
        }}
      >
        <h2 className={styles.login_title}>Login</h2>
        <label htmlFor="email-login" className={styles.email_label}>
          Email:
          <input
            onChange={(event) => {
              onChangeEmail(event).catch(() => {});
            }}
            name={Email}
            ref={refEmail}
            id="email-login"
            className={emailClass}
            type="text"
            placeholder="user@example.com"
            autoComplete="email"
            aria-invalid={errors.email || !emailState.isDirty ? 'true' : 'false'}
          />
        </label>
        {errors.email && (
          <span role="alert" className={styles.error}>
            {errors.email.message}
          </span>
        )}
        <div className={styles.password_block}>
          <label htmlFor="password-login" className={styles.password_label}>
            Password:
            <input
              onChange={(event) => {
                onChangePassword(event).catch(() => {});
              }}
              name={Password}
              ref={refPassword}
              id="password-login"
              className={passwordClass}
              type={revealPassword ? 'text' : 'password'}
              placeholder="password"
              autoComplete="current-password"
              aria-invalid={errors.password || !passwordState.isDirty ? 'true' : 'false'}
            />
          </label>
          <button
            type="button"
            className={revealPassword ? `${styles.reveal} ${styles.hidden}` : styles.reveal}
            aria-label="Reveal"
            onClick={() => setRevealPassword(!revealPassword)}
          />
        </div>
        {errors.password && (
          <span role="alert" className={styles.error}>
            {errors.password.message}
          </span>
        )}
        <CustomButton type="submit" isDisabled={!isValid}>
          Log in
        </CustomButton>
        <p className={styles.registry_link}>
          Haven’t registered yet? Sign up
          <Link to={ActionPaths.REGISTER} className={styles.link}>
            {` here!`}
          </Link>
        </p>
      </form>
    </main>
  );
}
