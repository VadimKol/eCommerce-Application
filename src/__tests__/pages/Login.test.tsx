import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { Login } from '@/pages/login/Login';
import { loginSchema } from '@/pages/login/login-schema';

describe('Testing login form', () => {
  test('Login test with good email and password', () => {
    expect(() => loginSchema.parse({ email: 'user@example.com', password: 'asdfasdfA1' })).not.toThrow();
  });

  test('Login test with invalid email and good password', () => {
    expect(() => loginSchema.parse({ email: 'user@examplecom', password: 'asdfasdfA1' })).toThrow();
  });

  test('Login test with good email and invalid password', () => {
    expect(() => loginSchema.parse({ email: 'user@example.com', password: 'asdfasdfA' })).toThrow();
  });

  test('Login test with invalid email and password', () => {
    expect(() => loginSchema.parse({ email: ' user@example.com ', password: 'asdfasdf1' })).toThrow();
  });

  it('renders correctly', () => {
    const { container } = render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );
    expect(container).toMatchSnapshot();
  });
});
