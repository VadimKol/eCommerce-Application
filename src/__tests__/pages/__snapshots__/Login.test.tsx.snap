// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing login form renders correctly 1`] = `
<div>
  <main
    class="main"
  >
    <form
      class="login"
    >
      <h2
        class="login_title"
      >
        Login
      </h2>
      <label
        class="email_label"
        for="email-login"
      >
        Email:
        <input
          aria-invalid="true"
          autocomplete="email"
          class="email"
          id="email-login"
          name="email"
          placeholder="user@example.com"
          type="text"
        />
      </label>
      <div
        class="password_block"
      >
        <label
          class="password_label"
          for="password-login"
        >
          Password:
          <input
            aria-invalid="true"
            autocomplete="current-password"
            class="password"
            id="password-login"
            name="password"
            placeholder="password"
            type="password"
          />
        </label>
        <button
          aria-label="Reveal"
          class="reveal"
          type="button"
        />
      </div>
      <button
        class="button button--primary"
        disabled=""
        id=""
        type="submit"
      >
        Log in
      </button>
      <p
        class="registry_link"
      >
        Havenâ€™t registered yet? Sign up
        <a
          class="link"
          href="/register"
        >
           here!
        </a>
      </p>
    </form>
  </main>
</div>
`;
